externals/kyuubi-spark-sql-engine/src/main/scala/org/apache/spark/kyuubi/SparkSQLLineageParseHelper.scala

```diff
@@ -0,0 +1,309 @@
+        }
+
+      case p if p.nodeName == "InsertIntoDataSourceCommand" =>
+        val logicalRelation = getPlanField[LogicalRelation]("logicalRelation")
+        val table = logicalRelation.catalogTable.map(_.qualifiedName).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          k.withName(s"$table.${k.name}") -> v
+        }
+
+      case p if p.nodeName == "InsertIntoHadoopFsRelationCommand" =>
+        val table =
+          getPlanField[Option[CatalogTable]]("catalogTable").map(_.qualifiedName).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          k.withName(s"$table.${k.name}") -> v
+        }
+
+      case p
+          if p.nodeName == "InsertIntoDataSourceDirCommand" ||
+            p.nodeName == "InsertIntoHiveDirCommand" =>
+        val dir =
+          getPlanField[CatalogStorageFormat]("storage").locationUri.map(_.toString).getOrElse("")
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          k.withName(s"`$dir`.${k.name}") -> v
+        }
+
+      case p if p.nodeName == "InsertIntoHiveTable" =>
+        val table = getPlanField[CatalogTable]("table").qualifiedName
+        extractColumnsLineage(getQuery, parentColumnsLineage).map { case (k, v) =>
+          k.withName(s"$table.${k.name}") -> v
+        }
+
+      case p if p.nodeName == "SaveIntoDataSourceCommand" =>
+        extractColumnsLineage(getQuery, parentColumnsLineage)
+
+      // For query
+      case p: Project =>
+        val nextColumnsLineage =
+          joinColumnsLineage(parentColumnsLineage, getSelectColumnLineage(p.projectList))
+        p.children.map(extractColumnsLineage(_, nextColumnsLineage)).reduce(mergeColumnsLineage)
+
+      case p: Aggregate =>
+        val nextColumnsLineage =
+          joinColumnsLineage(parentColumnsLineage, getSelectColumnLineage(p.aggregateExpressions))
+        p.children.map(extractColumnsLineage(_, nextColumnsLineage)).reduce(mergeColumnsLineage)
+
+      case p: Union =>
+        // merge all children in to one derivedColumns
+        val childrenUnion =
+          p.children.map(extractColumnsLineage(_, ListMap[Attribute, AttributeSet]())).map(
+            _.values).reduce {
+            (left, right) =>
+              left.zip(right).map(attr => attr._1 ++ attr._2)
+          }
+        val childrenColumnsLineage = ListMap(p.output.zip(childrenUnion): _*)
+        joinColumnsLineage(parentColumnsLineage, childrenColumnsLineage)
+
+      case p: LogicalRelation if p.catalogTable.nonEmpty =>
+        val tableName = p.catalogTable.get.qualifiedName
+        joinRelationColumnLineage(parentColumnsLineage, p.output, tableName)
+
+      case p: HiveTableRelation =>
+        val tableName = p.tableMeta.qualifiedName
+        joinRelationColumnLineage(parentColumnsLineage, p.output, tableName)
+
+      case p: DataSourceV2Relation =>
```

Comment:
For an optimzied plan, the v2 scan should always be `DataSourceV2ScanRelation`. We can remove this pattern